src/eager.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+// Think of Resources as File Descriptors. They are integers that are allocated
+// by the privlaged side of Deno to refer to various resources.  The simplest
+// example are standard file system files and stdio - but there will be other
+// resources added in the future that might not correspond to operating system
+// level File Descriptors. To avoid confusion we call them "resources" not "file
+// descriptors". This module implements a global resource table. Ops (AKA
+// handlers) look up resources by their integer id here.
+
+use resources::{self, Resource, RESOURCE_TABLE};
+use tokio_util;
+use tokio_write;
+
+use futures;
+use futures::future::Either;
+use futures::future::FutureResult;
+use std;
+use std::io::{Read, Write};
+use std::net::SocketAddr;
+use tokio;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_io;
+
+type EagerRead<R, T> =
+  Either<tokio_io::io::Read<R, T>, FutureResult<(R, T, usize), std::io::Error>>;
+
+#[cfg(windows)]
+#[allow(unused_mut)]
+pub fn read<T: AsMut<[u8]>>(
+  resource: Resource,
+  mut buf: T,
+) -> EagerRead<Resource, T> {
+  Either::A(tokio_io::io::read(resource, buf)).into()
+}
+
+#[cfg(not(windows))]
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd};
+
+#[cfg(not(windows))]
+fn read_tcp<T: AsMut<[u8]>>(
+  tcp_stream: &TcpStream,
+  resource: Resource,
+  mut buf: T,
+) -> EagerRead<Resource, T> {
+  // Unforunately we can't just call read() on tokio::net::TcpStream
+  let fd = (*tcp_stream).as_raw_fd();
+  let mut std_tcp_stream = unsafe { std::net::TcpStream::from_raw_fd(fd) };
+  let read_result = std_tcp_stream.read(buf.as_mut());
+  // std_tcp_stream will close when it gets dropped. Thus...
+  let _ = std_tcp_stream.into_raw_fd();
+  match read_result {
+    Ok(nread) => Either::B(futures::future::ok((resource, buf, nread))),
+    Err(err) => {
+      if err.kind() == std::io::ErrorKind::WouldBlock {
+        Either::A(tokio_io::io::read(resource, buf))
+      } else {
+        Either::B(futures::future::err(err))
+      }
+    }
+  }
+}
+
+// This is an optimization that Tokio should do.
+// Attempt to call read() on the main thread.
+#[cfg(not(windows))]
+pub fn read<T: AsMut<[u8]>>(
+  resource: Resource,
+  buf: T,
+) -> EagerRead<Resource, T> {
```

Comment:
Can you move this one into resources.rs 

ops -> resources -> eager

is the ideal dependency chain

Reply:
```
#[cfg(not(windows))]
pub fn read<T: AsMut<[u8]>>(
```

^-- that seems a little odd to move to a different file by itself.